ding/model/template/ebm.py

```diff
@@ -0,0 +1,497 @@
+            ebm: nn.Module, 
+            scheduler: BaseScheduler = None
+    ) -> torch.Tensor:
+        """
+        Run langevin MCMC for `self.iters` steps.
+        obs: (B, N, O), action: (B, N, A)
+        return: (B, N, A)
+        """
+        if not scheduler:
+            scheduler = MCMC.ExponentialScheduler(**self.stepsize_scheduler)
+        stepsize = scheduler.get_rate(-1)
+        for i in range(self.iters):
+            action = self._langevin_step(obs, action, stepsize, ebm)
+            stepsize = scheduler.get_rate(i)
+        return action
+
+    @no_ebm_grad()
+    def sample(self, obs: torch.Tensor, ebm: nn.Module) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Create tiled observations and sample counter-negatives for feeding to the InfoNCE objective.
+        obs: (B, O)
+        return: (B, N, O), (B, N, A)
+        """
+        obs, uniform_action_samples = self._sample(obs, self.train_samples)
+        if not self.use_langevin_negative_samples:
+            return obs, uniform_action_samples
+        langevin_action_samples = self._langevin_action_gives_obs(obs, uniform_action_samples, ebm)
+        return obs, langevin_action_samples
+
+    @no_ebm_grad()
+    def infer(self, obs: torch.Tensor, ebm: nn.Module) -> torch.Tensor:
+        """Optimize for the best action conditioned on the current observation.
+        obs: (B, O)
+        return: (B, A).
+        """
+        # (B, N, O), (B, N, A)
+        obs, uniform_action_samples = self._sample(obs, self.inference_samples)
+        action_samples = self._langevin_action_gives_obs(
+            obs, 
+            uniform_action_samples, 
+            ebm,
+        )
+
+        # Run a second optimization, a trick for more precise inference
+        if self.optimize_again:
+            self.again_stepsize_scheduler['num_steps'] = self.iters
+            action_samples = self._langevin_action_gives_obs(
+                obs, 
+                action_samples,
+                ebm,
+                scheduler=MCMC.PolynomialScheduler(**self.again_stepsize_scheduler),
+            )
+
+        # action_samples: B, N, A
+        return self._get_best_action_sample(obs, action_samples, ebm)
+
+
+
+@MODEL_REGISTRY.register('ebm')
+class EBM(nn.Module):
+    
+    def __init__(
+        self,
+        obs_shape: int,
+        action_shape: int,
+        hidden_size: int = 512,
+        hidden_layer_num: int = 4,
+        **kwargs,
+    ):
+        super().__init__()
+        input_size = obs_shape + action_shape 
+        self.net = nn.Sequential(
+            nn.Linear(input_size, hidden_size), 
+            nn.ReLU(),
+            RegressionHead(
+                hidden_size,
+                1,
+                hidden_layer_num,
+                final_tanh=False,
+            )
+        )
+
+    def forward(self, obs, action):
+        # obs: (B, N, O)
+        # action: (B, N, A)
+        # return: (B, N)
+        x = torch.concat([obs, action], -1)
```

Comment:
torch.concat does not work for torch==1.8.1
maybe torch.cat is a better substitute

Reply:
Got it. I will change it later.

Reply:
torch.cat